airbyte-integrations/connectors/source-surveymonkey/source_surveymonkey/streams.py

```diff
@@ -0,0 +1,257 @@
+        and returning an updated state object.
+        """
+        latest_state = latest_record.get(self.cursor_field)
+        current_state = current_stream_state.get(self.cursor_field) or latest_state
+        if current_state:
+            return {self.cursor_field: max(latest_state, current_state)}
+        return {}
+
+
+class Surveys(IncrementalSurveymonkeyStream):
+    """
+    Docs: https://developer.surveymonkey.com/api/v3/#surveys
+    A source for stream slices. It does not contain useful info itself.
+    """
+
+    cursor_field = "date_modified"
+
+    def path(self, **kwargs) -> str:
+        return "surveys"
+
+    def parse_response(self, response: requests.Response, stream_state: Mapping[str, Any], **kwargs) -> Iterable[Mapping]:
+        response_json = response.json()
+        if response_json.get("error"):
+            raise Exception(response_json.get("error"))
+        result = response_json.get(self.data_field, [])
+        for record in result:
+            substream = SurveyDetails(survey_id=record["id"], start_date=self._start_date, authenticator=self.authenticator)
+            child_record = substream.read_records(sync_mode=SyncMode.full_refresh)
+            yield from child_record
+
+
+class SurveyDetails(SurveymonkeyStream):
+    """
+    Actually stream above can just filter, sort and enumarate survey ids.
+    It does not contain the needed information. I will need it as stream slice source
+    in all of endpoints
+    """
+
+    def __init__(self, survey_id, start_date, **kwargs):
+        self.survey_id = survey_id
+        super().__init__(start_date=start_date, **kwargs)
+
+    def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
+        return f"surveys/{self.survey_id}/details"
+
+    def parse_response(self, response: requests.Response, stream_state: Mapping[str, Any], **kwargs) -> Iterable[Mapping]:
+        response_json = response.json()
+        if response_json.get("error"):
+            raise Exception(response_json.get("error"))
+        response_json.pop("pages", None)
+        yield response_json
+
+
+class SurveyPages(SurveymonkeyStream):
+    """should be filled from SurveyDetails"""
+
+    data_field = "pages"
+
+    def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
+        survey_id = stream_slice["survey_id"]
+        return f"surveys/{survey_id}/details"
+
+    def stream_slices(self, **kwargs):
+        survey_stream = Surveys(start_date=self._start_date, authenticator=self.authenticator)
+        for survey in survey_stream.read_records(sync_mode=SyncMode.full_refresh):
+            yield {"survey_id": survey["id"]}
+
+    def parse_response(self, response: requests.Response, stream_state: Mapping[str, Any], **kwargs) -> Iterable[Mapping]:
+        response_json = response.json()
+        data = response_json.get(self.data_field)
+        if response_json.get("error"):
```

Comment:
1. The CDK already throws an exception if the http header code is != 429 and is between 400 and 500. Is this still necessary in that case? 
2. If this is still necessary, why shouldn't we move it into the super class?